### DISCUSSION TITLE
Peter, please tell me.
About differential step.

### DISCUSSION THREAD

**Discussion by lex27087** [Peter-van-Tol/LiteX-CNC#93] [Q&A]
![Screenshot_20240221-222920.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/09cf0fc8-ea51-43d0-b12b-a6b0fc26031c)

![Screenshot_20240221-221101_1.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/88c7e01a-6412-4220-b42e-4e0a5aadc6b4)

**Comment by Peter-van-Tol** [comment]
You can create differential outputs in two ways:
- in the firmware with step_dir_differential as you describe;
- in hardware with a differential driver IC.

Because the 5A-75B has limited pins, I chose the second approach. I use the AM26LS31 as driver. See also the my HUB75HAT project.

**Comment by lex27087** [comment]
Thank you.

**Comment by lex27087** [comment]
Good day.
Here is the differential step.
Does everything seem normal?

**Comment by lex27087** [comment]
![1708587728815.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/63ce5329-7569-4c07-b90d-5204dee6b70c)

![IMG_20240222_104125_275.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/ec750f7a-5ef1-4ad8-a6b4-ae1a5632d09d)

**Reply by Peter-van-Tol** [reply]
This seems correct to me.

**Comment by lex27087** [comment]
Why do your teammates insist that such a signal is not correct?
Then why did you do it?
Thank you

**Reply by Peter-van-Tol** [reply]
Is this a question for me? There is nothing incorrect, there are only multiple ways how to do it.

**Comment by lex27087** [comment]
@Peter-van-Tol 
They say it should be like in the top picture.![IMG_20240222_135111_787.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/aa846ba2-0786-43b5-9319-5562b0eb4c05)

**Comment by GTCLive** [comment]
Just need to wear the summer shades' twisted-filter, and the worlds lines up!

![diff](https://github.com/Peter-van-Tol/LiteX-CNC/assets/91785220/894e3904-ab87-45c0-ab1a-7799ae7d608f)

**Comment by unromeo** [comment]
Just a quick question, if I may: is it ok to use differential signal just for the step pin, but normal for the dir pin ? I figure, the step signal is the one that would need more interference protection and has the faster rate, than the dir pin, so using less pins but hopefully not impact performance much ..

**Reply by Peter-van-Tol** [reply]
It depends on the drive you use whether yo will benefit from differential signals. I use Leadshine drives, which are designed to be driven single-ended. Reverse polarity on the opto-coupler inside the drive may lead to damage if the absolute rating is exceeded. I was looking for a discussion on this topic, but I cannot find it anymore.

Some drives are designed with differential inputs, such as the OMC Stepper Online iDM-series are designed with differential control. This example is a integrated stepper; i.e. the driver is integrated in the stepper, thus long cables for step and dir. For these drives I really see the benefit of differential signaling, because of those longer cables outside the cabinet.

To answer the question:
- if the drive is designed for differential signalling, I would recommend using it if the cables carrying step-dir signals are very long or outside the cabinet (integrated drives);
- if you want to use differential signalling, I recommend to use them for both step and dir. It's not only on the speed, but also signal integrity.
- if you are concerned about the number of pins on your FPGA, I can recommend using a differential line driver, such as the `26ls31` to create the differential signal.
